src/main/java/de/zalando/aruha/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -0,0 +1,366 @@
+            addTask(this::checkBatchTimeouts);
+        }
+    }
+
+    private void streamToOutput() {
+        if (!isCurrent()) {
+            return;
+        }
+        final long currentTimeMillis = System.currentTimeMillis();
+        int freeSlots = (int) getMessagesAllowedToSend();
+        SortedMap<Long, String> toSend;
+        for (final Map.Entry<Partition.PartitionKey, PartitionData> e : offsets.entrySet()) {
+            while (null != (toSend = e.getValue().takeEventsToStream(
+                    currentTimeMillis,
+                    Math.min(getParameters().batchLimitEvents, freeSlots),
+                    getParameters().batchTimeoutMillis))) {
+                flushData(e.getKey(), toSend);
+                this.sentEvents += toSend.size();
+                if (toSend.isEmpty()) {
+                    break;
+                }
+                freeSlots -= toSend.size();
+            }
+        }
+        pollPaused = getMessagesAllowedToSend() <= 0;
+        if (null != getParameters().batchKeepAliveIterations) {
+            final boolean allPartitionsReachedKeepAliveLimit = offsets.values().stream()
+                    .map(PartitionData::getKeepAlivesInARow)
+                    .allMatch(v -> v >= getParameters().batchKeepAliveIterations);
+            if (allPartitionsReachedKeepAliveLimit) {
+                shutdownGracefully();
+            }
+        }
+    }
+
+    private void flushData(final Partition.PartitionKey pk, final SortedMap<Long, String> data) {
+        final String evt = EventStream.createStreamEvent(
+                pk.partition,
+                String.valueOf(offsets.get(pk).getSentOffset()),
+                new ArrayList<>(data.values()),
+                Optional.empty());
+        try {
+            getOut().streamData(evt.getBytes(EventStream.UTF8));
+        } catch (final IOException e) {
+            LOG.error("Failed to write data to output. Session: " + getSessionId(), e);
+            shutdownGracefully();
+        }
+    }
+
+    @Override
+    public void onExit() {
+        if (null != topologyChangeSubscription) {
+            try {
+                topologyChangeSubscription.cancel();
+            } catch (final RuntimeException ex) {
+                LOG.warn("Failed to cancel topology subscription", ex);
+            } finally {
+                topologyChangeSubscription = null;
+                new HashSet<>(offsets.keySet()).forEach(this::removeFromStreaming);
+            }
+        }
+        if (null != kafkaConsumer) {
+            try {
+                kafkaConsumer.close();
+            } finally {
+                kafkaConsumer = null;
+            }
+        }
+    }
+
+    private void topologyChanged() {
+        if (!isCurrent()) {
+            return;
+        }
+        if (null != topologyChangeSubscription) {
+            topologyChangeSubscription.refresh();
+        }
+        reactNoTopologyChange();
+    }
+
+    private void reactNoTopologyChange() {
```

Comment:
This name is more about the context of execution and not what it does. I mean, by reading `reactNoTopologyChange` I know that it's doing something when a topology change happens but the change set is empty. But to understand what it does I have to read the code.